### PR TITLE
Fix cleaning send queue on restart

### LINKED ISSUE #17613: Hub handshake failed with error 'WebSocket is not in the OPEN state' SignalR WebSockets Angular client with AutoReconnect()
My application was first built on dotnetcore 2.2, then it was upgraded to dotnetcore 3.0 and it was working properly. But WithAutoReconnect() from Angular web app and all the default settings, gives following log:

```
[2019-12-05T12:10:14.538Z] Information: Reconnect attempt number 3 will start in 10000 ms.
zone.js:3372 OPTIONS http://127.0.0.1:9938/host/negotiate?clientIdentifier=b462ed0d-4df0-4f2d-82b1-4ad1ee72b09b net::ERR_CONNECTION_REFUSED
Utils.js:208 [2019-12-05T12:10:24.540Z] Debug: Starting connection with transfer format 'Text'.
Utils.js:208 [2019-12-05T12:10:24.540Z] Debug: Sending negotiation request: http://127.0.0.1:9938/host/negotiate?clientIdentifier=b462ed0d-4df0-4f2d-82b1-4ad1ee72b09b.
Utils.js:208 [2019-12-05T12:10:24.683Z] Debug: Selecting transport 'WebSockets'.
Utils.js:208 [2019-12-05T12:10:24.683Z] Trace: (WebSockets transport) Connecting.
Utils.js:204 [2019-12-05T12:10:24.723Z] Information: WebSocket connected to ws://127.0.0.1:9938/host?clientIdentifier=b462ed0d-4df0-4f2d-82b1-4ad1ee72b09b&id=KOrhcPrFdlHaeTzcJxZjlA.
Utils.js:208 [2019-12-05T12:10:24.724Z] Debug: The HttpConnection connected successfully.
Utils.js:208 [2019-12-05T12:10:24.724Z] Debug: Sending handshake request.
Utils.js:208 [2019-12-05T12:10:24.743Z] Debug: Hub handshake failed with error 'WebSocket is not in the OPEN state' during start(). Stopping HubConnection.
Utils.js:208 [2019-12-05T12:10:24.744Z] Trace: (WebSockets transport) socket closed.
Utils.js:208 [2019-12-05T12:10:24.744Z] Debug: HttpConnection.stopConnection(undefined) called while in state Disconnecting.
Utils.js:198 [2019-12-05T12:10:24.744Z] Error: Connection disconnected with error 'WebSocket is not in the OPEN state'.
Utils.js:208 [2019-12-05T12:10:24.745Z] Debug: HubConnection.connectionClosed(WebSocket is not in the OPEN state) called while in state Reconnecting.
Utils.js:204 [2019-12-05T12:10:24.745Z] Information: Reconnect attempt failed because of error 'WebSocket is not in the OPEN state'.
Utils.js:204 [2019-12-05T12:10:24.745Z] Information: Reconnect attempt number 4 will start in 30000 ms.
Utils.js:208 [2019-12-05T12:10:54.747Z] Debug: Starting connection with transfer format 'Text'.
Utils.js:208 [2019-12-05T12:10:54.748Z] Debug: Sending negotiation request: http://127.0.0.1:9938/host/negotiate?clientIdentifier=b462ed0d-4df0-4f2d-82b1-4ad1ee72b09b.
Utils.js:208 [2019-12-05T12:10:54.773Z] Debug: Selecting transport 'WebSockets'.
Utils.js:208 [2019-12-05T12:10:54.773Z] Trace: (WebSockets transport) Connecting.
Utils.js:204 [2019-12-05T12:10:54.780Z] Information: WebSocket connected to ws://127.0.0.1:9938/host?clientIdentifier=b462ed0d-4df0-4f2d-82b1-4ad1ee72b09b&id=JEu_f9AcM9fx4OE3SMWEVg.
Utils.js:208 [2019-12-05T12:10:54.780Z] Debug: The HttpConnection connected successfully.
Utils.js:208 [2019-12-05T12:10:54.781Z] Debug: Sending handshake request.
Utils.js:208 [2019-12-05T12:10:54.781Z] Trace: (WebSockets transport) sending data. String data of length 32.
Utils.js:204 [2019-12-05T12:10:54.782Z] Information: Using HubProtocol 'json'.
Utils.js:208 [2019-12-05T12:10:54.829Z] Trace: (WebSockets transport) data received. String data of length 3.
Utils.js:208 [2019-12-05T12:10:54.830Z] Debug: Server handshake complete.
Utils.js:204 [2019-12-05T12:10:54.831Z] Information: HubConnection reconnected successfully.
```

The server is disconnected and client tries to reconnect. When server goes up and connected; then the following steps occurred:
1. 3rd retry attempted and though "The HttpConnection connected successfully." but "Hub handshake failed with error 'WebSocket is not in the OPEN state' during start(). Stopping HubConnection" and (WebSockets transport) socket closed. HttpConnection.stopConnection(undefined) called while in state Disconnecting.
And as result "Error: Connection disconnected with error 'WebSocket is not in the OPEN state'." is logged in the console.
2. After 30sec 4th retry attempted and It will automatically reconnected. 
These steps are consistency reproducible.

Please help me to get rid of this error in between 3rd and 4th retry attempt.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/17613

Also slipped in a couple other minor fixups:
* Updated `karma-firefox-launcher` so that it would actually close the browser on test exit (I had dozens of firefox instances open from running the functional tests a bunch of times)
* Added an abort signal check in `NodeHttpClient` to fix a race between the LongPolling transport being closed and re-polling. (the race would only show up in tests since product code uses the `DefaultHttpClient` which already has the check)
* Changed some cleanup code in `WebSocketTransport` so we actually close the connection due to an "internal error". Such as throwing while parsing a message from the server.

### CODE REVIEW

File: src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js [hunk 1]
```diff
@@ -43,17 +43,17 @@ try {
     }
 
     // We use the launchers themselves to figure out if the browser exists. It's a bit sneaky, but it works.
-    tryAddBrowser("ChromeHeadlessNoSandbox", new ChromeHeadlessBrowser(() => { }, {}));
-    tryAddBrowser("ChromiumHeadlessIgnoreCert", new ChromiumHeadlessBrowser(() => { }, {}));
-    if (!tryAddBrowser("FirefoxHeadless", new FirefoxHeadlessBrowser(0, () => { }, {}))) {
-      tryAddBrowser("FirefoxDeveloperHeadless", new FirefoxDeveloperHeadlessBrowser(0, () => { }, {}));
+    tryAddBrowser("ChromeHeadlessNoSandbox", ChromeHeadlessBrowser.prototype);
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js]: This is part of updating the firefox launcher. We shouldn't have been instantiating these classes to check if the browsers exists.

File: src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js [hunk 2]
```diff
@@ -43,17 +43,17 @@ try {
...
+    if (!tryAddBrowser("FirefoxHeadless", FirefoxHeadlessBrowser.prototype)) {
+      tryAddBrowser("FirefoxDeveloperHeadless", FirefoxDeveloperHeadlessBrowser.prototype);
     }
 
     // We need to receive an argument from the caller, but globals don't seem to work, so we use an environment variable.
     if (process.env.ASPNETCORE_SIGNALR_TEST_ALL_BROWSERS === "true") {
-      tryAddBrowser("Edge", new EdgeBrowser(() => { }, { create() { } }));
-      tryAddBrowser("IE", new IEBrowser(() => { }, { create() { } }, {}));
-      tryAddBrowser("Safari", new SafariBrowser(() => { }, {}));
+      tryAddBrowser("Edge", EdgeBrowser.prototype);
```
Comment (jkotalik) [inline src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js]: Which edge is this testing? Legacy?

File: src/SignalR/clients/ts/FunctionalTests/yarn.lock [hunk 3]
```diff
@@ -43,21 +43,6 @@
   resolved "https://registry.yarnpkg.com/@types/node/-/node-9.6.49.tgz#ab4df6e505db088882c8ce5417ae0bc8cbb7a8a6"
   integrity sha512-YY0Okyn4QXC4ugJI+Kng5iWjK8A6eIHiQVaGIhJkyn0YL6Iqo0E0tBC8BuhvYcBK87vykBijM5FtMnCqaa5anA==
 
-"@types/strip-bom@^3.0.0":
```
Comment (jkotalik) [inline src/SignalR/clients/ts/FunctionalTests/yarn.lock]: I assume this is expected?

File: src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js (hunk 2 above)
Comment (anurse) [reply src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js]: TBH we need to do some work to understand what Karma does locally and what Sauce Labs will do in the cloud. I'll file a bug for that.

For now, I'd expect yes, it's Edge Legacy.

File: src/SignalR/clients/ts/signalr/tests/TestWebSocket.ts [hunk 4]
```diff
@@ -27,7 +28,10 @@ export class TestWebSocket {
     }
 
     public get onopen(): (this: WebSocket, evt: Event) => any {
-        return this._onopen!;
+        return (e) => {
+            this._onopen!(e);
+            this.readyState = this.OPEN;
```
Comment (jkotalik) [inline src/SignalR/clients/ts/signalr/tests/TestWebSocket.ts]: Where is readyState used?

**Review by jkotalik — APPROVED** [review]

File: src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js (hunk 2 above)
Comment (anurse) [reply src/SignalR/clients/ts/FunctionalTests/scripts/karma.local.conf.js]: https://github.com/dotnet/aspnetcore/issues/18524

File: src/SignalR/clients/ts/FunctionalTests/yarn.lock (hunk 3 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/FunctionalTests/yarn.lock]: Are you referring to the file changing? Yes, I updated dependencies so this file gets updated.

File: src/SignalR/clients/ts/signalr/tests/TestWebSocket.ts (hunk 4 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/tests/TestWebSocket.ts]: Used by the WebSocketTransport

File: src/SignalR/clients/ts/signalr/src/WebSocketTransport.ts [hunk 5]
```diff
@@ -139,12 +138,7 @@ export class WebSocketTransport implements ITransport {
 
     public stop(): Promise<void> {
         if (this.webSocket) {
-            // Clear websocket handlers because we are considering the socket closed now
-            this.webSocket.onclose = () => {};
-            this.webSocket.onmessage = () => {};
-            this.webSocket.onerror = () => {};
-            this.webSocket.close();
-            this.webSocket = undefined;
+            this.cleanup();
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/WebSocketTransport.ts]: Nit: close() also calls cleanup(), and close() is called immediately after this. I'd just remove cleanup() and put the logic directly in close().

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 6]
```diff
@@ -497,6 +488,13 @@ export class HttpConnection implements IConnection {
             this.logger.log(LogLevel.Information, "Connection disconnected.");
         }
 
+        if (this.sendQueue) {
+            this.sendQueue.stop().catch((e) => {
+                this.logger.log(LogLevel.Error, `TransportSendQueue.stop() threw error '${e}'.`);
+            });
+            this.sendQueue = undefined;
+        }
+
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: While we're cleaning up random things, can we move the `this.connectionStarted = false;` above the `if (this.onclose && this.connectionStarted) {`? I wrote it that way, but don't know why. It shouldn't be dependent on whether the onclose callback was hooked. Thankfully, HubConnection always does hook the onclose callback, but still.

**Review by halter73 — APPROVED** [review]

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 6 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I moved the `if (this.onclose)` check instead. We don't want to run the `onclose` function if the connection never started.

Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: > We don't want to run the onclose function if the connection never started.

Yeah. That's the whole point of that logic. We don't need to check if connectionStarted is true in order to set it to false. Both ways work, but it does seem a little weird to put the if block inside the try block.

**Pull request merged**